### PR TITLE
Avoid allocations in Display for OsStr and Path

### PR DESCRIPTION
**Pull request by stepancheg**
#38879

### CODE REVIEW

File: src/libstd/path.rs [hunk 1]
```diff
@@ -2281,8 +2281,8 @@ impl AsRef<OsStr> for Path {
 
 #[stable(feature = "rust1", since = "1.0.0")]
 impl fmt::Debug for Path {
-    fn fmt(&self, formatter: &mut fmt::Formatter) -> Result<(), fmt::Error> {
-        self.inner.fmt(formatter)
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Debug::fmt(&self.inner, formatter)
```
Comment (nikomatsakis) [inline src/libstd/path.rs]: is this just general cleanup, or was this required for some reason?

File: src/libstd_unicode/lossy.rs [hunk 2]
```diff
@@ -0,0 +1,154 @@
+use core::str as core_str;
+use core::fmt;
+use core::fmt::Write;
+use char;
+
+/// Iterator over lossy UTF-8 string
+#[unstable(feature = "str_internals", issue = "0")]
+pub struct Utf8LossyIter<'a> {
```
Comment (nikomatsakis) [inline src/libstd_unicode/lossy.rs]: can we write some more unit tests for this? it seems like quite a big chunk of code, and one for which it'd be easy to write some targeted unit tests...

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
This looks nice! The main thing seems to be that we should add some targeted unit tests for the new iterator-- I presume, perhaps wrongly, that there are existing tests for the `Display` code in paths...

File: src/libstd/path.rs (hunk 1 above)
Comment (stepancheg) [inline src/libstd/path.rs]: I've made several similar changes for these reasons:

* it is easier to visually check that correct trait is called (`Debug` or `Display`)
* to make sure code on Windows and Redox (e. g. `os_str::Slice`) works correctly. I do not have access to Windows machine, so it's easier to write code uniformly (single `fmt` import and explicit trait names) than wait for build system job completion
* and after all to unify all implementations in stack `os_str`, `OsStr`, `Path`

File: src/libcollections/string.rs [hunk 3]
```diff
@@ -535,111 +535,37 @@ impl String {
...
+        } else {
+            return Cow::Borrowed("");
+        };
 
-        let mut res = String::with_capacity(total);
+        const REPLACEMENT: &'static [u8] = b"\xEF\xBF\xBD"; // U+FFFD in UTF-8
 
-        if i > 0 {
-            unsafe { res.as_mut_vec().extend_from_slice(&v[..i]) };
+        let mut res = String::with_capacity(v.len());
```
Comment (arthurprs) [inline src/libcollections/string.rs]: We know it'll be larget than v.len right? Maybe reserve some extra, maybe v.len() * 3 / 2?

Comment (stepancheg) [inline src/libcollections/string.rs]: Probably. If yes, this should be done in separate PR. I've just kept this part as it was before my patch.

**Comment by stepancheg** [comment]
I've added tests (copied from `String::from_utf8_lossy` tests).

**Comment by stepancheg** [comment]
Some tests broken, wait a bit...

**Comment by stepancheg** [comment]
There's a tidy check error:

```
tidy error: /Users/yozh/devel/left/rust/src/libstd/ffi/os_str.rs:321: different `since` than before
tidy error: /Users/yozh/devel/left/rust/src/libstd/ffi/os_str.rs:685: different `since` than before
```

There were no `Display` for `OsStr` and `OsString` which are need to implement `Display` for `Path` and which I've added. What should I do?

I've added

```
#[stable(feature = "rust1", since = "1.20.0")]
```

which is apparently incorrect.

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
Going to pass the buck here. I'm not that familiar with the attributes that need to be set.

r? @alexcrichton

File: src/libstd/ffi/os_str.rs [hunk 4]
```diff
@@ -312,12 +312,19 @@ impl Default for OsString {
...
-impl Debug for OsString {
-    fn fmt(&self, formatter: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+impl fmt::Debug for OsString {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
         fmt::Debug::fmt(&**self, formatter)
     }
 }
 
+#[stable(feature = "rust1", since = "1.20.0")]
+impl fmt::Display for OsString {
```
Comment (alexcrichton) [inline src/libstd/ffi/os_str.rs]: This did not exist before and is a new stable feature being added in this PR, is that intentional? It was the libs team intention that this impl did *not* exist.

File: src/libcore/fmt/mod.rs [hunk 5]
```diff
@@ -1512,23 +1512,30 @@ impl Display for bool {
     }
 }
 
+#[doc(hidden)]
+#[unstable(feature = "str_internals", issue = "0")]
+pub fn escape_str(s: &str, f: &mut Formatter) -> Result {
```
Comment (alexcrichton) [inline src/libcore/fmt/mod.rs]: Can this avoid being exposed? It seems like a reasonable amount that's not too bad to just copy directly into the new iterator.

File: src/libstd/ffi/os_str.rs (hunk 4 above)
Comment (stepancheg) [inline src/libstd/ffi/os_str.rs]: `Display` for `OsString` seems to be perfectly reasonable for me. You may need to "display" an OS string like you "display" the usual string.

Anywhay, if it should exist, how should I mute `tidy`?

If it shouldn't exist, how should I keep an implementation, but hide it from public API? `#[doc(hidden)]` and `#[unstable(feature = "str_internals", issue = "0")]`?

Comment (alexcrichton) [inline src/libstd/ffi/os_str.rs]: Whether or not it's reasonable it is intentionally excluded and should not be added in this PR. If you'd like to add it then that should be done through a separate PR or an RFC.

This shouldn't exist, which means it should be deleted (along with most other `Display` impls added here)

**Comment by stepancheg** [comment]
Updated the PR.

* Removed public `Display` implementations (kept implementations for private types in `os_str`)
* Copied implementation of `Debug` for `str` to `Debug` for `Utf8Lossy` instead of extraction of `escape_str` function

**Pull request merged**